### ISSUE TITLE
Prevent data loss with automated backups for files and unsaved changes

### ISSUE THREAD

**Issue by kumarbalap** [Microsoft/vscode#1161]
Windows 10 is half baked one IE and now this editor.

I was surprised with the built in extensions, but I'm feeling terrible after experiencing this issue.

I was working with this editor, and suddenly my system became very slow and I'm forced to shut down my system. After the restart my current working file (see the attachment), got changed in to some other encoding. I lost my 3 hours of work :(.
- Bala
  [global.txt](https://github.com/Microsoft/vscode/files/57803/global.txt)

**Comment by egamma** [comment]
>Windows 10 is half baked one IE and now this editor.

Sorry about the bad experience. 

What was the encoding of the file before you killed Code?

@bpasero when I try to open the attached file I get "... the file uses an unsupported encoding"

**Comment by bpasero** [comment]
@kumarbalap so did you work for 3 hours without saving or saving frequently? Did you save the file before shutting down or was auto save enabled?

We have heard reports from users that vs code can fail to write the contents of the file to disk in certain cases where the OS shuts down and in that case the contents are not correctly written to the file, leaving the file corrupt.

**Comment by kumarbalap** [comment]
Thanks for the replies.

I took a backup in the morning, after that I was saving it frequently, but not committed. So I cudn't recover the content.

**Comment by kumarbalap** [comment]
@bpasero , as you said, it might be because of OS and auto save was not enabled by default.
But this is the first time I'm experiencing this issue, I have used many editors before.

Thank you

**Comment by bpasero** [comment]
@kumarbalap it would still be interesting if the corruption happened because of you saving the file or something else. When you shutdown the OS, was the file dirty in the editor or saved? Did the OS shutdown maybe while you saved?

**Comment by kumarbalap** [comment]
@bpasero system suddenly hanged, so I did forced shutdown. I have a strong feeling that the editor made the system slow. The file was fine till the restart, when I opened after the restart it's corrupted. I did saved the file several times before this happened. So whichever the files not saved, got corrupted. One php and js file.

**Comment by bpasero** [comment]
@kumarbalap so it happened for more than just one file?

**Comment by kumarbalap** [comment]
Yes @bpasero

**Comment by bpasero** [comment]
@kumarbalap any chance that maybe your disk was getting full and you ran out of space?

**Comment by bpasero** [comment]
Did some investigation. When we open a file for saving, node will first truncate the file and then write to it. There is a small chance that the file ends up empty if the process quits before the write has happened. I was able to reproduce this in a small sample (writing to a large file and then terminating the process).

However I was not able to reproduce the 0-bytes being written to the file, so it must be something else.

There is also some discussion over at https://github.com/atom/atom/issues/3158 that indicates issues when the disk is running out of space.

We could provide an option to enable more atomic saving (write to a tmp file and then replace the file with one fs call), but it is also worth mentioning that other popular editors chose to disable this by default, because e.g. external file watchers might get confused that a file is constantly being replaced instead of changed.

**Comment by rotemdan** [comment]
Using the 'write to a copy, then rename' technique to achieve more atomic writes is a useful and viable solution but I wanted to also mention some additional, alternative approaches based on the file system itself. These would provide a more 'thorough' solution to the issue, and would also apply to other text editors and external tools (compilers, code generators, refactoring tools etc.).

First, the file contains 4436 bytes of the 'NUL' (0) character, which basically means it got replaced by a block of zeros, there is no special encoding to it.

I had this happen to me before (same 0 data, for a code file of approximately the same length), but with the 'full' visual studio. This appeared after I recovered from a sudden power failure. Fortunately I had a shortcut that triggers 'one-click' backups (which also immediately syncs the resulting compressed file to google drive). I initiated it several seconds before this happened, so I managed to recover the data without any loss.

I can speculate a part of the reason this happened because the file system was set to use ['writeback mode/write caching mode'](http://thewindowsclub.thewindowsclubco.netdna-cdn.com/wp-content/uploads/2013/01/disk-caching.jpg) (this is the default setting in windows), so although the new file content was probably cached in memory (possibly in full) the OS only managed to flush the metadata, but not the data itself by the time the system crashed. The metadata, which included some pointer(s) to a physical area of the disk (that probably happened to be initially filled with zeroes), and the file size, were written, but the data itself was not.

In Linux the default setting for ext3/ext4 is to use what is called 'ordered mode', it is a journaling mode where the file system takes care to only flush metadata after the data is physically flushed to disk (a form of a write barrier). As far as I'm aware NTFS does not support ordered mode/write barriers. However, I'm not sure if ordered mode would have been sufficient to prevent data loss here. Using a 'full' journaling mode (not supported by NTFS), which is explained below, would prevent failures of this sort as the data is written twice (once to the journal, and then to its physical location). Another alternative is using an allocate-on-write file system like btrfs (linux) or ReFS (windows - details on how to enable and use it below).

If I were using linux I would have probably created a separate `ext4` partition and mount it with `data=journal` (or alternatively use `btrfs` instead) and keep all my code and development files there (no need to set for the whole system to run in this mode as it would significantly reduce performance, `data=ordered` is probably sufficient and is [very good in performance](https://fsi-viewer.blogspot.com/2011/10/filesystem-benchmarks-part-i.html) for general usage).

From [the ext4 documentation](https://www.kernel.org/doc/Documentation/filesystems/ext4.txt):
```
Data Mode
=========
There are 3 different data modes:

* writeback mode
In data=writeback mode, ext4 does not journal data at all.  This mode provides
a similar level of journaling as that of XFS, JFS, and ReiserFS in its default
mode - metadata journaling.  A crash+recovery can cause incorrect data to
appear in files which were written shortly before the crash.  This mode will
typically provide the best ext4 performance.

* ordered mode
In data=ordered mode, ext4 only officially journals metadata, but it logically
groups metadata information related to data changes with the data blocks into a
single unit called a transaction.  When it's time to write the new metadata
out to disk, the associated data blocks are written first.  In general,
this mode performs slightly slower than writeback but significantly faster than journal mode.

* journal mode
data=journal mode provides full data and metadata journaling.  All new data is
written to the journal first, and then to its final location.
In the event of a crash, the journal can be replayed, bringing both data and
metadata into a consistent state.  This mode is the slowest except when data
needs to be read from and written to disk at the same time where it
outperforms all others modes.  Enabling this mode will disable delayed
allocation and O_DIRECT support.
```

In windows 8 or later, it is possible to [enable single volumes to be formatted with ReFS instead of NTFS](https://www.jmedved.com/2015/03/refs-on-windows-81/). ReFS is an 'allocate-on-write' file system and should provide a higher level of reliability than NTFS (both metadata and data blocks are never overwritten, so the possiblity of data loss due to power or program failures is reduced). However, to get the maximum level of 'resiliency' (mostly data recovery in case of corruption) it is currently recommended to be used with a multiple disk array rather than a single one (and also cannot be used as a boot partition). 

From http://theithollow.com/2014/01/13/microsofts-resilient-file-system-refs/ 
```
Torn Writes

One of the new things that ReFS helps with is a “torn write”.  In NTFS the file 
system overwrites metadata and data structures during a modify operation.  During 
this modify operation data is read off the disk and then new data is written on top 
of the old data, becoming the new data.  Torn writes occur when this process can’t 
complete fully.  Think about what could happen if a power failure occurred during 
the write operation.  Maybe some of the new data is there, and some of the old data 
is available, but neither are consistent and there is no way to roll this back.  
With ReFS new data will be written to a different location and when the write is 
complete, the file system then references the new location as the up to date data. 
This process is known as “Copy on Write”.  Now during a power failure, if the write
 doesn’t complete fully the original data is still consistent.
```

As an experiment, I went ahead and created a single disk 8GB ReFS partition exclusively for storing source code and similar data. I will need some time to evaluate what kind of benefit this setup would have but I believe it should provide a meaningful improvement in data safety and crash-resilience. One disadvantage is that by default ReFS has a cluster size of 64kb which may be too large for source files, which tend to be relatively small (e.g. after copying 5593 files to the new volume, where many are small, windows reported the directory to have a total of 584MB of allocated size but only 272MB of actual size). I read somewhere it might be possible to change this but it is not mentioned as an option for the `format` command. Also, the link I provided has instructions on how to enable integrity checking (`format .. -I:enable``), though in a case of a failure it would not be possible to fix the problem without some form of a mirrored array (this is handled when the volume is created using [Storage Spaces](https://en.wikipedia.org/wiki/Features_new_to_Windows_8#Storage_Spaces), I believe).

**Comment by rotemdan** [comment]
There's a very good summary of various crash resistence properties of common file system in this paper:

[All File Systems Are Not Created Equal: On the Complexity of Crafting Crash-Consistent Applications](https://www.usenix.org/conference/osdi14/technical-sessions/presentation/pillai)
 
![Persistence property chart](http://i.imgur.com/rmyaY25.png)

Unfortunately `ReFS` is not included in the paper but I believe it should have properties similar to `btrfs`.

For the most common case of overwriting source files, which usually do not exceed 64kb (a single ReFS block/cluster), the write should be atomic in ReFS. For multi-block appends (not overwrites), it depends. The paper describes these writes as "Multi-block prefix append", and states:

> Current file systems do not provide atomic multi-block
appends; appends can be broken down into multiple operations.
However, most file systems seemingly guarantee
that some prefix of the data written (e.g., the first 10
blocks of a larger append) will be appended atomically.

So I speculate that with large block sizes, appending to files up to, say, half a megabyte or so may be atomic in btrfs/ReFS, but I can't know for sure. It depends on how the writes are batched and ordered.

For atomic _overwrites_ beyond the block size (usually 4kb to 64kb), using the sequence:
1. Create a secondary file and write the new content to it (`fopen`, `fwrite`).
2. Flush the new file to the physical media (`fsync`).
3. Rename the new copy to the old copy (`rename`).

Should work (though the `fsync` part is important for it to provide full reliability).

**Comment by bpasero** [comment]
Good summaries, thanks. One gotcha with rename is that at least on Windows you cannot rename across different drive letters. So it is good to keep the temporary file outside of the workspace to prevent the issues of file watchers (external tools) picking it up, but you cannot just store it in the default tmp location if the target workspace is on a different drive.

**Comment by rotemdan** [comment]
@bpasero 

I went into the node/libuv source code to see exactly how `rename` is implemented:
It seems that in windows it is [using the `MoveFileExW` API call](https://github.com/libuv/libuv/blob/bddd6a8489e4c8cf47841de6f05becd99fc06f3e/src/win/fs.c#L1215) (with the `MOVEFILE_REPLACE_EXISTING` flag), documented [here](https://msdn.microsoft.com/en-us/library/aa365240%28v=vs.85%29.aspx), so at least in windows, it may actually support moving between different volumes, though that is highly unlikely to be atomic (there's also the [`MoveFileTransacted`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365241%28v=vs.85%29.aspx) but I don't think it is supported by node). In any case, in unix (this includes mac I believe) it simply uses the [`rename` call](https://github.com/libuv/libuv/blob/bddd6a8489e4c8cf47841de6f05becd99fc06f3e/src/unix/fs.c#L816) which probably doesn't work across volumes.

One thing that's interesting is how exactly would the rename/move be 'transparently' supported in the editor itself. I mean, since a new file is written and then moved on top the old file, an existing file descriptor would still point to the old file (the old file would only be truly deleted once the file descriptor is released). I'm guessing that the editor is opening and closing the file on each save? (or using the `writeFile`, `readFile` functions in the node API?).

Another thing is how to handle multiple pending saves to a file. Since `fsync` may take a while there could be a situation (especially when auto-save is enabled) where multiple pending overwrites are happening at the same time. So I guess the temporary file would need to be timestamped like `MySourceFile.ts.1450956207523_temp` (I'm using a millisecond epoch timestamp here since multiple saves could happen during the same second).

Another issue to be tested is how watcher tools interact with these quick deletion and moves, do they consider them as file updates or a new file creation? (this may vary between watchers). Also, if an external program happens to have a live file descriptor to the file, even for a fraction of a second while an atomic save is in progress, then it will not be notified of any changes to it and a write to the file may result in an error or data loss (the file descriptor they hold is basically referencing a 'phantom' file which at that point isn't referenced by any directory).

**Comment by rotemdan** [comment]
An alternative to atomic file overwrites is to simply keep a timestamped backup for files that are frequently being edited, in some dedicated location (say: in the application data folder). There are should be at least one valid backup of the file at any given time. For example:

```
For MySourceFile.ts:

MySourceFile.ts.backup_1450966677534
```

After every new backup is created and flushed to disk with `fsync` (an atomic write/rename/flush sequence would be needed here), the oldest one of the existing backups is deleted.

One issue is, since I'm assuming the preferred constraint is how to store and match the full path of the file to the correct backup, even if the backup is stored in a different directory (or even a different volume). That's actually quite easy, instead of using a complicated scheme to store file metadata somewhere, simply hash the full path of the file (including the drive, directory path and file name) and name the backup by that hash:

```
For 'C:\My Dev Directory\My Project\MySourceFile.ts' the backup filename would be:

ee1eb030124f7aa68fa740b5ce89b4060396b3a7_1450966677534
```
(the hex string is the SHA1 of the complete path of the file)

To restore a backup the IDE could add an option when right clicking the file in the editor and then selecting something like 'Restore from latest backup' (if a backup is unavailable the option could be grayed or not appear at all). Then the user would be notified of the exact date and time of the backup and asked whether to proceed (this action should preferably be undo-able).

**Comment by bpasero** [comment]
Yes, we do use `fs.writeFile()` API of node to save a file and we never leave any file handles open. We also avoid any parallel access to files when writing, so I do not think this would be an issue. 

Using the atomic write strategy is not a 100% guarantee to never loose data, but it reduces the chance of a failing write. 

Watchers would receive a delete and create event and some are smart enough to convert this into a change event, but there is no guarantee. Code itself will convert to change events, but other watchers might not and get confused.

I think doing a backup of the file in some other location is the only good solution. The question is wether to enable this by default or not. I know that Atom is now optionally providing a setting to enable backups of files.

**Comment by rotemdan** [comment]
I edited my comment but I would post this again:

The automated backups could even be applied to unsaved files, and the editor can then compare the file date of the actual file to its backup. If the backup is newer it may ask something like 'a unsaved version of the file has been automatically stored, would you like to load it?' (I'm aware auto-save should cover some of the scenarios where this is useful, but I guess not everyone use it. In any case, the backup approach provides an advantage on the atomic save approach for this matter).

**Comment by bpasero** [comment]
Yes I would totally do backups of dirty files as well as it would be the fix for #1106. In fact I will close that one as duplicate of this one and rename this one.

**Comment by Tyriar** [comment]
Duplicate https://github.com/Microsoft/vscode/issues/12327
